## Analysing Datasets from 2014 to 2024-nov


Daily Prices of Market Yard Commodities in Telangana
Agriculture
Organisation: Agricultural Marketing and Co-operation Department

Data available from 01-01-2014 to 30-11-2024

This dataset contains information on the daily prices of all the commodities across all the market yards in the state of Telangana 

Contains the below columns: | DDate | AmcCode | AmcName | YardCode | YardName | CommCode | CommName | VarityCode | VarityName | Arrivals | Minimum | Maximum | Modal |
 
Abbreviation: 
 
DDate: Date 
AmcCode: Agricultural Market Committee Code 
AmcName: Agricultural Market Committee Name 
YardCode: Agricultural Market Yard Code 
YardName: Agricultural Market Yard Name 
CommCode: Commodity Code 
CommName: Commodity Name 
VarityCode: Commodity Variety Code 
VarityName: Commodity Variety Name 
Arrivals: Quantity in Quintals (Qtls) 
Minimum: Minimum Price per Quintal 
Maximum: Maximum Price per Quintal
Modal: Modal Price is the price per quintal at which most transactions took place

In [10]:
import pandas as pd

# Read the CSV files
data1 = pd.read_csv('commoditypricesTsDatasetFrom2014to2024nov/day_prices_between_01-11-2024_30-11-2024.csv')
data2 = pd.read_csv('commoditypricesTsDatasetFrom2014to2024nov/day_prices_between_01-02-2023_28-02-2023.csv')
data3 = pd.read_csv('commoditypricesTsDatasetFrom2014to2024nov/day_prices_between_01-08-2021_31-08-2021.csv')
data4 = pd.read_csv('commoditypricesTsDatasetFrom2014to2024nov/day_prices_between_01-01-2017_31-12-2017.csv')
data5 = pd.read_csv('commoditypricesTsDatasetFrom2014to2024nov/day_prices_between_01-01-2014_31-12-2014.csv')
# Store the DataFrames in a dictionary with keys as their names
data_dict = {
    'data1': data1,
    'data2': data2,
    'data3': data3,
    'data4': data4,
    'data5': data5
}

# Loop through the dictionary and print info for each DataFrame
for i in range(1, 6):
    data_key = f'data{i}'  # Create the variable name (e.g., 'data1')
    print(f"Information for {data_key}:")
    data_dict[data_key].info()  # Access and print the info of the DataFrame
    print("#" * 40)


Information for data1:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8615 entries, 0 to 8614
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   DDate         8615 non-null   object 
 1   AmcCode       8615 non-null   int64  
 2   AmcName       8615 non-null   object 
 3   YardCode      8615 non-null   int64  
 4   YardName      8615 non-null   object 
 5   CommCode      8615 non-null   int64  
 6   CommName      8615 non-null   object 
 7   VarityCode    8615 non-null   int64  
 8   VarityName    8615 non-null   object 
 9   ProgArrivals  8615 non-null   float64
 10  Arrivals      8615 non-null   float64
 11  Minimum       8615 non-null   float64
 12  Maximum       8615 non-null   float64
 13  Model         8615 non-null   float64
 14  Valuation     8615 non-null   float64
 15  MarketFee     8615 non-null   float64
dtypes: float64(7), int64(4), object(5)
memory usage: 1.1+ MB
##################################

###### we observe that all columns if every dataset is not same.

## Merging all the datasets

In [14]:
import os
import pandas as pd

# Path to the folder containing CSV files
folder_path = 'commoditypricesTsDatasetFrom2014to2024nov'

# List to hold DataFrames
dataframes = []

# Loop through all files in the folder
for file in os.listdir(folder_path):
    if file.endswith('.csv'):  # Check if the file is a CSV
        file_path = os.path.join(folder_path, file)
        df = pd.read_csv(file_path)  # Read CSV into DataFrame
        dataframes.append(df)  # Append to the list

# Ensure all DataFrames have the same columns by using an outer join
all_columns = pd.concat([df.columns.to_frame().T for df in dataframes], ignore_index=True).stack().unique()
aligned_dataframes = [df.reindex(columns=all_columns) for df in dataframes]

# Concatenate all DataFrames
merged_df = pd.concat(aligned_dataframes, ignore_index=True)

# Save the merged DataFrame to a new CSV file
output_path = 'dataset1/merged_file.csv'
merged_df.to_csv(output_path, index=False)

print(f"Merged file saved to {output_path}")


Merged file saved to dataset1/merged_file.csv


In [2]:
import pandas as pd
df=pd.read_csv('dataset1/merged_file.csv')
df.info()
df.describe()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 711341 entries, 0 to 711340
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   DDate         711341 non-null  object 
 1   AmcCode       711341 non-null  int64  
 2   AmcName       711341 non-null  object 
 3   YardCode      711341 non-null  int64  
 4   YardName      711341 non-null  object 
 5   CommCode      711341 non-null  int64  
 6   CommName      711341 non-null  object 
 7   VarityCode    711341 non-null  int64  
 8   VarityName    711341 non-null  object 
 9   Arrivals      711341 non-null  float64
 10  Minimum       711341 non-null  float64
 11  Maximum       711341 non-null  float64
 12  Model         711341 non-null  float64
 13  ProgArrivals  392920 non-null  float64
 14  Valuation     392920 non-null  float64
 15  MarketFee     392920 non-null  float64
dtypes: float64(7), int64(4), object(5)
memory usage: 86.8+ MB


DDate  AmcCode     AmcName  YardCode    YardName  \
0  2014-01-02 00:00:00+05:30      246    Warangal       886    Warangal   
1  2014-01-02 00:00:00+05:30      221  Jammikunta        36  Jammikunta   
2  2014-01-03 00:00:00+05:30      246    Warangal       886    Warangal   
3  2014-01-04 00:00:00+05:30      246    Warangal       886    Warangal   
4  2014-01-06 00:00:00+05:30      246    Warangal       886    Warangal   

   CommCode CommName  VarityCode     VarityName  Arrivals  Minimum  Maximum  \
0         1    Paddy           1         Common     637.0   1310.0   1330.0   
1         1    Paddy         240  Common (1001)     161.0   1300.0   1310.0   
2         1    Paddy           1         Common     287.0   1310.0   1330.0   
3         1    Paddy           1         Common     329.0   1310.0   1330.0   
4         1    Paddy           1         Common     149.0   1500.0   1675.0   

    Model  ProgArrivals  Valuation  MarketFee  
0  1320.0           NaN        NaN        NaN  
1  1310.0           NaN        NaN        NaN  
2  1320.0           NaN        NaN        NaN  
3  1320.0           NaN        NaN        NaN  
4  1600.0           NaN        NaN        NaN

In [3]:
df['DDate'] = pd.to_datetime(df['DDate'])

###### In columns ProgArrivals,Valuation,MarketFee there are 392920 non-null values and rest  711341-392920 may contain some value

## Data Analysis

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 711341 entries, 0 to 711340
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype                    
---  ------        --------------   -----                    
 0   DDate         711341 non-null  datetime64[ns, UTC+05:30]
 1   AmcCode       711341 non-null  int64                    
 2   AmcName       711341 non-null  object                   
 3   YardCode      711341 non-null  int64                    
 4   YardName      711341 non-null  object                   
 5   CommCode      711341 non-null  int64                    
 6   CommName      711341 non-null  object                   
 7   VarityCode    711341 non-null  int64                    
 8   VarityName    711341 non-null  object                   
 9   Arrivals      711341 non-null  float64                  
 10  Minimum       711341 non-null  float64                  
 11  Maximum       711341 non-null  float64                  
 12  Model         71

In [5]:
# weighted_avg_price = (df['frequency'] * df['price']).sum() / df['frequency'].sum()
df_2014 = df[df['DDate'].dt.year == 2014]
df_2014.info()
weighted_sum = df_2014.groupby('CommName').agg(
    weighted_sum=('Arrivals', lambda x: (x * df_2014.loc[x.index, 'Model']).sum()),
    total_arrivals=('Arrivals', 'sum')
)

print(weighted_sum)

<class 'pandas.core.frame.DataFrame'>
Index: 3117 entries, 0 to 3116
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype                    
---  ------        --------------  -----                    
 0   DDate         3117 non-null   datetime64[ns, UTC+05:30]
 1   AmcCode       3117 non-null   int64                    
 2   AmcName       3117 non-null   object                   
 3   YardCode      3117 non-null   int64                    
 4   YardName      3117 non-null   object                   
 5   CommCode      3117 non-null   int64                    
 6   CommName      3117 non-null   object                   
 7   VarityCode    3117 non-null   int64                    
 8   VarityName    3117 non-null   object                   
 9   Arrivals      3117 non-null   float64                  
 10  Minimum       3117 non-null   float64                  
 11  Maximum       3117 non-null   float64                  
 12  Model         3117 non-null   float64  

In [6]:
pd.set_option('display.max_rows', None)
value_counts=df['CommName'].value_counts()
filtered_df = df[df['CommName'].isin(value_counts[value_counts >= 10000].index)]
filtered_df['CommName'].unique()


array(['Paddy', 'Maize', 'Cotton', 'Red Gram', 'Green Gram',
       'Groundnut pods', 'Chillies(Dry)', 'Tumeric', 'Onions', 'Potato',
       'Cabbage', 'Brinjal', 'Tomato', 'Cauliflower', 'Ladys Finger',
       'Green Chillies', 'Carrot', 'French Beans', 'Cluster Beans',
       'Bottle Gourd', 'Field Beans', 'Bitter Gourd', 'Ribbed Gourd',
       'Capsicum', 'Kheera', 'Cucumber', 'Donda'], dtype=object)

#### removing all the row or CommName whose count is less than 10,000

In [7]:
df.shape


(711341, 16)

In [8]:
filtered_df.shape

(533483, 16)

In [9]:
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 533483 entries, 0 to 711340
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype                    
---  ------        --------------   -----                    
 0   DDate         533483 non-null  datetime64[ns, UTC+05:30]
 1   AmcCode       533483 non-null  int64                    
 2   AmcName       533483 non-null  object                   
 3   YardCode      533483 non-null  int64                    
 4   YardName      533483 non-null  object                   
 5   CommCode      533483 non-null  int64                    
 6   CommName      533483 non-null  object                   
 7   VarityCode    533483 non-null  int64                    
 8   VarityName    533483 non-null  object                   
 9   Arrivals      533483 non-null  float64                  
 10  Minimum       533483 non-null  float64                  
 11  Maximum       533483 non-null  float64                  
 12  Model         533483 

In [10]:
dropcategory=['ProgArrivals','Valuation','MarketFee']
filtered_df.drop(labels=dropcategory,axis=1,inplace=True)
filtered_df.info()

C:\Users\abdul\AppData\Local\Temp\ipykernel_1780\3129733443.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df.drop(labels=dropcategory,axis=1,inplace=True)


<class 'pandas.core.frame.DataFrame'>
Index: 533483 entries, 0 to 711340
Data columns (total 13 columns):
 #   Column      Non-Null Count   Dtype                    
---  ------      --------------   -----                    
 0   DDate       533483 non-null  datetime64[ns, UTC+05:30]
 1   AmcCode     533483 non-null  int64                    
 2   AmcName     533483 non-null  object                   
 3   YardCode    533483 non-null  int64                    
 4   YardName    533483 non-null  object                   
 5   CommCode    533483 non-null  int64                    
 6   CommName    533483 non-null  object                   
 7   VarityCode  533483 non-null  int64                    
 8   VarityName  533483 non-null  object                   
 9   Arrivals    533483 non-null  float64                  
 10  Minimum     533483 non-null  float64                  
 11  Maximum     533483 non-null  float64                  
 12  Model       533483 non-null  float64             

In [11]:
new_df = filtered_df[['DDate', 'CommName', 'Model']]
new_df.info()
df=new_df

<class 'pandas.core.frame.DataFrame'>
Index: 533483 entries, 0 to 711340
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype                    
---  ------    --------------   -----                    
 0   DDate     533483 non-null  datetime64[ns, UTC+05:30]
 1   CommName  533483 non-null  object                   
 2   Model     533483 non-null  float64                  
dtypes: datetime64[ns, UTC+05:30](1), float64(1), object(1)
memory usage: 16.3+ MB


In [12]:
from sklearn.preprocessing import LabelEncoder
encoder=LabelEncoder()
df['encoded']=encoder.fit_transform(df['CommName'])
df['Year'] = pd.to_datetime(df['DDate']).dt.year
df['Month'] = pd.to_datetime(df['DDate']).dt.month
df['Day'] = pd.to_datetime(df['DDate']).dt.day
df['DayOfWeek'] = pd.to_datetime(df['DDate']).dt.dayofweek
df['Quarter'] = pd.to_datetime(df['DDate']).dt.quarter



C:\Users\abdul\AppData\Local\Temp\ipykernel_1780\1846114779.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['encoded']=encoder.fit_transform(df['CommName'])
C:\Users\abdul\AppData\Local\Temp\ipykernel_1780\1846114779.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Year'] = pd.to_datetime(df['DDate']).dt.year
C:\Users\abdul\AppData\Local\Temp\ipykernel_1780\1846114779.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,co

In [18]:
from sklearn.model_selection import train_test_split
target=['Model']
features=['encoded','Year','Month','Day','DayOfWeek','Quarter']
X=df[features]
y=df[target]
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [20]:
X.info()
y.info()

<class 'pandas.core.frame.DataFrame'>
Index: 533483 entries, 0 to 711340
Data columns (total 6 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   encoded    533483 non-null  int64
 1   Year       533483 non-null  int32
 2   Month      533483 non-null  int32
 3   Day        533483 non-null  int32
 4   DayOfWeek  533483 non-null  int32
 5   Quarter    533483 non-null  int32
dtypes: int32(5), int64(1)
memory usage: 18.3 MB
<class 'pandas.core.frame.DataFrame'>
Index: 533483 entries, 0 to 711340
Data columns (total 1 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Model   533483 non-null  float64
dtypes: float64(1)
memory usage: 8.1 MB


In [22]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.fit_transform(X_test)

In [23]:
from sklearn.linear_model import LinearRegression
model=LinearRegression()

In [24]:
model.fit(X_train_scaled,y_train)

LinearRegression()

In [26]:
print(y_test.head())
print(y_pred[:5])


         Model
648911  2600.0
272448  4800.0
699083  3500.0
75270   4811.0
655704  8660.0
[[-1.04256751e+53]
 [ 2.43465007e+52]
 [-1.27193132e+53]
 [ 5.49076631e+52]
 [-4.81285583e+52]]


In [28]:
from sklearn.metrics import mean_squared_error
y_pred=model.predict(X_test_scaled)
mse=mean_squared_error(y_test,y_pred)
print(mse)

8.678962770366803e+105


In [36]:
input_data = {
    'encoded': 'Tomato',  # Categorical feature
    'Year': 2020,
    'Month': 12,
    'Day': 15,
    'DayOfWeek': 1,
    'Quarter': 4
}
input_data['encoded'] = encoder.transform([input_data['encoded']])[0]
input_df = pd.DataFrame([input_data])
input_scaled = scaler.transform(input_df)
prediction = model.predict(input_scaled)

original_prediction = scaler.inverse_transform(prediction.reshape(-1, 1))
decoded_prediction = encoder.inverse_transform(prediction)
print("Prediction:", original_prediction)

ValueError: non-broadcastable output operand with shape (1,1) doesn't match the broadcast shape (1,6)